### ISSUE TITLE
Neo4j 3.5.1 Optional Match Returning Null

### ISSUE THREAD

**Issue by bonesyo** [neo4j/neo4j#12122]
Version: 3.5.1
OS: Debian 8

Queries containing OPTIONAL MATCH are failing. They were working well in 3.4.9, but are returning null in 3.5.1. Below is an example of a query that returns null for "s":

```
MATCH (r:Room { id: "1234 })
OPTIONAL MATCH (r)-[hs:HAS_SKU]->(s:SKU {sku: "123456"})
RETURN r, s
```

If I change the OPTIONAL MATCH to a standard MATCH, it works as expected.
```
MATCH (r:Room { id: "1234 })
MATCH (r)-[hs:HAS_SKU]->(s:SKU {sku: "123456"})
RETURN r, s
```

^^ works^^
This is happening in both our production API and in the web browser. There are indexes on both room(id) and sku(sku).

**Comment by Hunterness** [comment]
Hi, thanks for reporting, we will look into it.

**Comment by bonesyo** [comment]
@Hunterness Thank you! This was a major breaking issue for our app. So please let me know if you need any more info.

**Comment by InverseFalcon** [comment]
@bonesyo Does that entire pattern in the OPTIONAL MATCH exist in the graph? If it does then this is likely a bug. If the pattern doesn't exist (whether the node for `s` exists or not) then the OPTIONAL MATCH is behaving correctly.

**Comment by bonesyo** [comment]
@InverseFalcon It does exist, like I said if I run the second query where I replace the OPTIONAL MATCH with MATCH, it runs correctly, which means that the entire pattern exists.

**Comment by InverseFalcon** [comment]
@bonesyo Thanks, I think I misread your queries initially. Possible bug, but please run a [consistency check](https://neo4j.com/docs/operations-manual/current/tools/consistency-checker/) on your db so we can see if anything else is in play.

**Comment by craigtaverner** [comment]
Is it possible to provide a small database or a Cypher query to generate the data, where this problem can be reproduced, because the behaviour of optional match is well tested, and so we suspect there must be something very specific about this case that triggers the bug, and we would need data to help reproduce it.

**Comment by bonesyo** [comment]
@craigtaverner how can I reach you to send over a backup of my development database, as well as a query that is not returning the expected results?

**Comment by craigtaverner** [comment]
@bonesyo Perhaps you can send it to me on the neo4j-users.slack.com, my user there is 'craig.taverner'.

**Comment by Lojjs** [comment]
@bonesyo Thanks for the additional data that you sent. I have investigated a bit today and it turns out that the issue you are seeing is not about OPTIONAL MATCH after all, but is a bug in the slotted runtime with aggregations (in your case `collect`) with many grouping keys. It is the same bug that was reported here: https://github.com/neo4j/neo4j/issues/12113, so it is already fixed.

To get the bug fix you can upgrade to 3.5.2. Meanwhile, you can run your query with the community runtime (might have performance implications) by prepending it with `CYPHER runtime=interpreted`.

Best regards Louise, Neo4j Cypher team

**Comment by bonesyo** [comment]
@Lojjs To be clear the initial issue I posted here was something that did happen in our production database, however I was not able to re-create it in my development database. So I still think something funky is going on.
